pkg/scheduler/objects/queue.go

```diff
@@ -1292,3 +1304,67 @@ func (sq *Queue) String() string {
 	return fmt.Sprintf("{QueuePath: %s, State: %s, StateTime: %x, MaxResource: %s}",
 		sq.QueuePath, sq.stateMachine.Current(), sq.stateTime, sq.maxResource)
 }
+
+func (sq *Queue) incRunningApps() {
+	if sq == nil {
+		return
+	}
+	if sq.parent != nil {
+		sq.parent.incRunningApps()
+	}
+	sq.internalIncRunningApps()
+}
+
+func (sq *Queue) internalIncRunningApps() {
+	sq.Lock()
+	defer sq.Unlock()
+	sq.runningApps++
+}
+
+func (sq *Queue) decRunningApps() {
+	if sq == nil {
+		return
+	}
+	if sq.parent != nil {
+		sq.parent.decRunningApps()
+	}
+	sq.internalDecRunningApps()
+}
+
+func (sq *Queue) internalDecRunningApps() {
+	sq.Lock()
+	defer sq.Unlock()
+	sq.runningApps--
+}
+
+func (sq *Queue) canRun() bool {
+	if sq == nil {
+		return false
+	}
+	ok := true
+	if sq.parent != nil {
+		ok = sq.parent.canRun()
+	}
+	return sq.internalCanRun(ok)
+}
+
+func (sq *Queue) internalCanRun(ok bool) bool {
+	sq.RLock()
+	defer sq.RUnlock()
+	if sq.maxRunningApps == 0 {
+		return true && ok
+	}
+	return ok && sq.runningApps < sq.maxRunningApps
+}
```

Comment:
I had to really look twice to grasp what was going on with the `ok` flag. Isn't this a much simpler flow and faster, as we fail fast if a parent has failed:
```
func (sq *Queue) canRun() bool {
	if sq == nil {
		return false
	}
	// check the parent(s)
	if sq.parent != nil {
		ok := sq.parent.canRun()
		if !ok {
			return false
		}
	}
	// check this queue only if all parents allowed it
	sq.RLock()
	defer sq.RUnlock()
	if sq.maxRunningApps == 0 {
		return true
	}
	return sq.runningApps < sq.maxRunningApps
}
```

Reply:
same reason as before, also I updated logic in canRun https://github.com/apache/yunikorn-core/pull/435

Reply:
@craigcondit Actually I removed canRun() method, this file is outdated. 
Current canRun logic is inside this method: https://github.com/apache/yunikorn-core/pull/429/commits/030c1b44a69ec8302fc566f57af7489b26d5cc0e#diff-27632d48eb925e150a33bc92370ceaa66c31048018d11ca7a53a0b50ab7250acR1345

Please review and lmk if you have any comments.